### PR TITLE
Feature/issue 1736

### PR DESCRIPTION
**Pull request by yantonglu**
Fixed issue#1736, discontinued `cogent3.maths.special.permutations(), cogent3.maths.special.permutations_exact(), cogent3.maths.special.ln_permutations()` and removed corresponding test `test_maths/test_stats/test_special.py::SpecialTests::test_permutations(),test_permutations_errors(),test_permutations_float(),test_permutations_range(),test_permutations_exact(),test_ln_permutations()`

### CODE REVIEW

File: src/cogent3/maths/stats/special.py [hunk 1]
```diff
@@ -58,45 +62,28 @@ def one_minus_exp(x):
...
-
-    Note: no error checking (expects to be called through permutations())
-    """
-    return lgam(n + 1) - lgam(n - k + 1)
+@c3warn.deprecated_callable(
+    version="2024.9", reason="use scipy.special.perm() instead", is_discontinued=True
+)
+def ln_permutations(n, k):  # pragma: no cover
+    """being removed"""
+    return log(scipy.special.perm(n, k, exact=False))
```
Comment (GavinHuttley) [inline src/cogent3/maths/stats/special.py:86]: The point here is not to replace our function with that from Scipy; rather, just use the decorator as indicated. Revert the change within the function to the original.

File: src/cogent3/maths/stats/special.py [hunk 2]
```diff
@@ -58,45 +62,28 @@ def one_minus_exp(x):
...
-    if min(n, k) < 20 and isinstance(n, int) and isinstance(k, int):
-        return permutations_exact(n, k)
-    else:
-        return exp(ln_permutations(n, k))
+@c3warn.deprecated_callable(
+    version="2024.9", reason="use scipy.special.perm() instead", is_discontinued=True
+)
+def permutations(n, k):  # pragma: no cover
+    """being removed"""
+    return scipy.special.perm(n, k, exact=False)
```
Comment (GavinHuttley) [inline src/cogent3/maths/stats/special.py:70]: The point here is not to replace our function with that from Scipy; rather, just use the decorator as indicated. Revert the change within the function to the original.

File: src/cogent3/maths/stats/special.py [hunk 3]
```diff
@@ -58,45 +62,28 @@ def one_minus_exp(x):
...
 
 
-def permutations_exact(n, k):
-    """Calculates permutations by integer division.
+@c3warn.deprecated_callable(
+    version="2024.9", reason="use scipy.special.perm() instead", is_discontinued=True
+)
+def permutations_exact(n, k):  # pragma: no cover
+    """being removed"""
+    return scipy.special.perm(n, k, exact=True)
```
Comment (GavinHuttley) [inline src/cogent3/maths/stats/special.py:78]: The point here is not to replace our function with that from Scipy; rather, just use the decorator as indicated. Revert the change within the function to the original.

File: src/cogent3/maths/stats/special.py [hunk 4]
```diff
@@ -2,8 +2,12 @@
 (c) Stephen L. Moshier 1984, 1995.
 """
 
+import scipy
```
Comment (GavinHuttley) [inline src/cogent3/maths/stats/special.py:5]: Delete this import. The point here is not to replace our function with that from Scipy; rather, just use the decorator and tell users where to go to find the replacement.

**Review by GavinHuttley — CHANGES_REQUESTED** [review]
Thanks for the contribution!

When marking a function to be discontinued we limit our changes to applying the decorator and giving the reason (including referring users to a replacement in another project). In other words, we do not actually change our code within the function since we are going to be deleting it.

**Comment by yantonglu** [comment]
OK, I'll revert them.

**Comment by yantonglu** [comment]
I have reverted the changes within functions, please review them again. Thanks!

**Review by GavinHuttley — APPROVED** [review]
awesome job @yantonglu , thanks for the contribution!

**Comment by GavinHuttley** [comment]
fixes #1736

**Pull request merged**